Optimize replace into to do a blind write

oleksandr-kachan opened the pull request:
Reference Patch: https://github.com/facebook/mysql-5.6/commit/f14c64cf950

---------- https://github.com/facebook/mysql-5.6/commit/f14c64cf950 ----------

Summary:
This is a diff based of https://github.com/facebook/mysql-5.6/pull/913. The initial pull request was submitted by Zhang Yuan from Alibaba. I am trying to continue work on this.

Approach:
During sql_insert, if the statement is a 'replace into' statement, then we set a boolean in the lex to indicate that primary key violations can be ignored. This is done only for tables that have an explicitly defined primary key, does not have any triggers or secondary keys.

If the lex flag indicates that this is a blind replace into, then engine ignores duplicate primary key errors. This in turn ensures that the sql layer does not have to do a delete followed by a insert.

If the lex flag is set, then Write_rows_log_event sets a bit in its flag indicating that this was an optimized 'replace into' statement execution. This is carried over to the slave and the slave sets the lex flag (in Write_rows_log_event::write_row()) before invoking the handler's write_row methods.

On the slave, if blind replace bit is set for a write rows event and it is possible to do a blind replace, we will proceed with it.
If for some reason (such as slaves having triggers or secondary keys), we cannot do a blind replace then this will fall back to doing either an update or a delete+insert. Note that the binlog will not have a before image (since on master this is just an insert), but we can still do the regular update or delete+insert.

Originally Reviewed By: abhinav04sharma

fbshipit-source-id: be7cfcb2f70

bhatvinay committed:
Optimize replace into to do a blind write

Reference Patch: https://github.com/facebook/mysql-5.6/commit/f14c64cf950

---------- https://github.com/facebook/mysql-5.6/commit/f14c64cf950 ----------

Summary:
This is a diff based of https://github.com/facebook/mysql-5.6/pull/913. The initial pull request was submitted by Zhang Yuan from Alibaba. I am trying to continue work on this.

Approach:
During sql_insert, if the statement is a 'replace into' statement, then we set a boolean in the lex to indicate that primary key violations can be ignored. This is done only for tables that have an explicitly defined primary key, does not have any triggers or secondary keys.

If the lex flag indicates that this is a blind replace into, then engine ignores duplicate primary key errors. This in turn ensures that the sql layer does not have to do a delete followed by a insert.

If the lex flag is set, then Write_rows_log_event sets a bit in its flag indicating that this was an optimized 'replace into' statement execution. This is carried over to the slave and the slave sets the lex flag (in Write_rows_log_event::write_row()) before invoking the handler's write_row methods.

On the slave, if blind replace bit is set for a write rows event and it is possible to do a blind replace, we will proceed with it.
If for some reason (such as slaves having triggers or secondary keys), we cannot do a blind replace then this will fall back to doing either an update or a delete+insert. Note that the binlog will not have a before image (since on master this is just an insert), but we can still do the regular update or delete+insert.

Originally Reviewed By: abhinav04sharma

fbshipit-source-id: be7cfcb2f70

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine-master.opt:
```diff
@@ -0,0 +1 @@
+--enable_blind_replace
```
For command-line options general convention is to use dashes(`-`) instead of underscores(`_`)
```suggestion
--enable-blind-replace
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine-slave.opt:
```diff
@@ -0,0 +1 @@
+--enable_blind_replace
```
Likewise

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+--source include/have_log_bin.inc
+--source include/have_binlog_format_row.inc
+--source include/master-slave.inc
```
Move `--source include/master-slave.inc` to the very end of the `--source` section (after all lightweight `have_xxx.inc`)
Currently, because of the `have_debug.inc` this test will be skipped in Release configuration but this will happen only after a long-lasting master-slave initialization process.

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+
+# Create a table in master with rocksdb as storage engine
+--echo Create a table in master with rocskdb as storage engine
+connection master;
```
```suggestion
--source include/rpl_connection_master.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+
+# Change table to have innodb as storage engine
+-- echo Table is altered in slave to be in innodb storage engine
+connection slave;
```
```suggestion
--source include/rpl_connection_slave.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+connection master;
+create table t1 (c1 int primary key, c2 int) engine=rocksdb;
+select table_name, engine from information_schema.tables where table_name = 't1';
+sync_slave_with_master;
```
`sync_slave_with_master` is not considered reliable use the following include file instead
```suggestion
--source include/sync_slave_sql_with_master.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+alter table t1 engine=innodb;
+select table_name, engine from information_schema.tables where table_name = 't1';
+
+connection master;
```
```suggestion
--source include/rpl_connection_master.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+select * from t1;
+sync_slave_with_master;
+
+connection slave;
```
```suggestion
--source include/rpl_connection_slave.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+
+# Issue a replace into in master. This should be a blind insert
+--echo replace into should be a blind replace in master and a regular replace in slave
+connection master;
```
```suggestion
--source include/rpl_connection_master.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+--source include/show_binlog_events.inc
+sync_slave_with_master;
+
+connection slave;
```
```suggestion
--source include/rpl_connection_slave.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+--source include/show_binlog_events.inc
+
+# Cleanup
+connection master;
```
```suggestion
--source include/rpl_connection_master.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+insert into t1 values(2, 2);
+insert into t1 values(3, 3);
+select * from t1;
+sync_slave_with_master;
```
```suggestion
--source include/sync_slave_sql_with_master.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+# binlog event should indicate a Write_rows event in master and Update_rows event in slave
+select * from t1;
+--source include/show_binlog_events.inc
+sync_slave_with_master;
```
```suggestion
--source include/sync_slave_sql_with_master.inc
```

percona-ysorokin commented on mysql-test/suite/rpl/t/optimized_replace_mixed_engine.test:
```diff
@@ -0,0 +1,55 @@
+# Cleanup
+connection master;
+drop table t1;
+sync_slave_with_master;
```
```suggestion
--source include/sync_slave_sql_with_master.inc
```

percona-ysorokin commented on sql/sys_vars.cc:
```diff
@@ -7622,3 +7622,11 @@ static Sys_var_bool Sys_tx_read_only(
     READ_ONLY SESSION_VAR(transaction_read_only), NO_CMD_LINE, DEFAULT(0),
     NO_MUTEX_GUARD, NOT_IN_BINLOG);
+
+static Sys_var_bool Sys_enable_blind_replace(
```
In 8.0 `Sys_var_bool` has 2 more additional parameters. Please, double-check

Commit:
Fix dereferencing of NULL IOStatsContext pointer

Summary:
In case ROCKSDB_SUPPORT_THREAD_LOCAL is not defined and
rocksdb_perf_context_level variable is set to some value
an attempt to insert a row into a table with engine=rocksdb
leads to SIGSEGV. This happens in Rdb_io_perf::start as a result
of an attempt to dereference NULL pointer to IOStatsContext.

percona-ysorokin reviewed (approved):
LGTM
